index.bs

```diff
@@ -8064,6 +8119,124 @@ JavaScript Array values.
 </div>
 
 
+<h4 id="js-async-iterable">Async iterable — async iterable&lt;|T|&gt;</h4>
+
+IDL <a lt="async iterable type">async iterable&lt;|T|&gt;</a> values are represented by a JavaScript
+object and a JavaScript [=Iterator=] record.
+
+<div id="js-to-async-iterable" algorithm="convert a JavaScript value to async iterable">
+    A JavaScript value |V| is [=converted to an IDL value|converted=]
+    to an IDL <a lt="async iterable type">async iterable&lt;<var ignore>T</var>&gt;</a> value as follows:
+
+    1.  If <a abstract-op>Type</a>(|V|) is not Object,
+        [=JavaScript/throw=] a <l spec=ecmascript>{{TypeError}}</l>.
+    1.  Let |iteratorRecord| be [=?=] <a abstract-op>GetIterator</a>(|V|, async).
+    1.  Return |iteratorRecord|.
+</div>
+
+<div id="async-iterable-to-js" algorithm="convert an async iterable to a JavaScript value">
+    An IDL <a lt="async iterable type">async iterable&lt;|<var ignore>T</var>&gt;</a> value is
+    [=converted to a JavaScript value|converted=] to a JavaScript object as follows:
+
+    1.  Return the JavaScript object that represents the same async iterable as the IDL value.
+</div>
```

Comment:
When would we need this? If I understand correctly, IDL interfaces should always add an async iterable declaration to themselves, rather than returning an async iterable from one of its operations.

Also, if the async iterable value can only have been created by converting it from a JavaScript value, is this value still usable when converting it back to a JavaScript value? The previous algorithm calls `GetIterator(V, async)`, which for a `ReadableStream` *locks the stream*. When we return the `ReadableStream` back to JavaScript, it'll still be locked by IDL and JavaScript won't be able to async-iterate the stream.

Would it be an option for us to just *disallow* converting an IDL async iterable type back to a JavaScript value?

Reply:
Hmm, maybe the real problem is that the IDL machinery should not be calling `GetIterator()` from within the conversion logic in the first place. 🤔 For example: if an IDL operation accepts *two* `async iterable<T>` values as parameters, and `GetIterator()` succeeds for the first value but fails for the second value (e.g. because it's a locked `ReadableStream`), then the operation errors immediately with a `TypeError` but the first value is now permanently "locked".

I think we should make "convert from a JavaScript object" *not* call `GetIterator()`:
* Make an IDL async iterable *only* represent a JavaScript object.
* Add a separate type for an "IDL async iterator", which is represented by a JavaScript `Iterator` record.
* Add an "open an async iterable" algorithm, which takes an IDL async iterable, calls `GetIterator()` and returns it as an IDL async iterator.
* Refactor the "get the next value of an async iterable" and "close an async iterable" algorithms to work on an IDL async iterator instead.

Thoughts?

Reply:
@MattiasBuelens I have attempted to implement this in [5533589](https://github.com/whatwg/webidl/pull/1397/commits/5533589f036f80ab6dbcff63ba1c5bad8dba03d2)